module/jobqueue/README.md

```diff
@@ -0,0 +1,70 @@
+
+So the goal for the jobqueue system are:
+1. guarantee each job (i.e. finalized block) will be processed eventually
+2. in the event of a crash failure, the jobqueue state is persisted and workers can be rescheduled so that no job is skipped.
+3. allow concurrent processing of multiple jobs
+4. the number of concurrent workers is configurable so that the node won't get overwhelmed when too many jobs are created (i.e. too many blocks are finalized in a short period of time)
+
+## JobQueue components
+To achieve the above goal, the jobqueue system contains the following components/interfaces:
+1. A `Jobs` module to find jobs by job index
+2. A `storage.ConsumerProgress` to store job processing progress
+3. A `Worker` module to process jobs and report job completion.
+4. A `Consumer` that orchestrates the job processing by finding new jobs, creating workers for each job using the above modules, and managing job processing status internally.
+
+### Using module.Jobs to find jobs
+There is no JobProducer in jobqueue design. Job queue assumes each job can be indexed by a uint64 value, just like each finalized block (or sealed block) can be indexed by block height.
+
+Let's just call this uint64 value "Job Index" or index.
+
+So if we iterate through each index from low to high, and find each job by index, then we are able to iterate through each job.
+
+Therefore modules.Job interface abstracts it into a method: `AtIndex`.
+
+`AtIndex` method returns the job at any given index.
+
+Job consumer relies on the modules.Jobs to find jobs. However, modules.Jobs doesn't provide a way to notify as soon as a new job is available. So it's consumer's job to keep track of the values returned by module.Jobs's `Head` method and find jobs that are new.
+
+### Using Check method to notify job consumer for checking new jobs
+Job consumer provides the `Check` method for users to notify new jobs available.
+
+Once called, job consumer will iterate through each height with the `AtIndex` method. It stops when one of the following condition is true:
+1. no job was found at a index
+2. no more workers to work on them
+3. max search ahead is reached (will explain later)
+
+`Check` method is concurrent safe, meaning even if job consumer is notified concurrently about new jobs available, job consumer will check at most once to find new jobs.
+
+Whenever a worker finishes a job, job consumer will also call `Check` internally.
+
+### Storing job consuming progress in storage.ConsumerProgress
+Job consumer stores the last processed job index in `storage.ConsumerProgress`, so that on startup, the job consumer can read the last processed job index from storage and compare with the last available job index from `module.Jobs`'s `Head` method to resume job processing.
+
+This ensures each job will be processed at least once. Note: given the at least once execution, the `Worker` should gracefully handle duplicate runs of the same job.
```

Comment:
You may want to mention that on job execution failure process will die with `Fatal`